src/nodecache.h

```diff
@@ -1,52 +1,67 @@
 #pragma once
 
-#include <vector>
 #include <string>
+#include <map>
 #include "node.h"
 #include "memory.h"
+#include "assert.h"
+#include "printutils.h"
 
 /*!
 	Caches string values per node based on the node.index().
 	The node index guaranteed to be unique per node tree since the index is reset
 	every time a new tree is generated.
 */
+
+typedef std::pair<long, long> IndexPair;
+
 class NodeCache
 {
 public:
-  NodeCache() { }
-  virtual ~NodeCache() { }
-
-	bool contains(const AbstractNode &node) const {
-		if (this->cache.size() > node.index()) return this->cache[node.index()].get();
-		return false;
-	}
-
-  /*! Returns a reference to the cached string copy. NB! don't rely on
-	 *  this reference to be valid for long - if the cache is resized
-	 *  internally, existing values are lost.  */
-  const std::string & operator[](const AbstractNode &node) const {
-    if (this->cache.size() > node.index()) return *this->cache[node.index()];
-    else return this->nullvalue;
-  }
-
-  /*! Returns a reference to the cached string copy. NB! don't rely on
-	 *  this reference to be valid for long - if the cache is resized
-	 *  internally, existing values are lost. */
-  const std::string &insert(const class AbstractNode &node, const std::string & value) {
-    if (this->cache.size() <= node.index()) this->cache.resize(node.index() + 1);
-		this->cache[node.index()].reset(new std::string(value));
-    return *this->cache[node.index()];
-  }
-
-  void remove(const class AbstractNode &node) {
-    if (this->cache.size() > node.index()) this->cache[node.index()].reset();
-  }
-
-	void clear() {
-		this->cache.clear();
-	}
+    NodeCache() { }
+    virtual ~NodeCache() { cache.clear(); }
+
+    bool contains(const AbstractNode &node) const {
+        auto result = cache.find(node.index()); 
+        return result != cache.end() && result->second.second >= 0L;
+    }
+
+    const std::string operator[](const AbstractNode &node) const {
```

Comment:
The const qualifier for the return value doesn't add anything as it's passed by value anyway.